go/ephemeral/team_ek.go

```diff
@@ -0,0 +1,413 @@
+package ephemeral
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/client/go/teams"
+)
+
+type TeamEKSeed keybase1.Bytes32
+
+func newTeamEphemeralSeed() (seed TeamEKSeed, err error) {
+	randomSeed, err := makeNewRandomSeed()
+	if err != nil {
+		return seed, err
+	}
+	return TeamEKSeed(randomSeed), nil
+}
+
+func newTeamEKSeedFromBytes(b []byte) (s TeamEKSeed, err error) {
+	seed, err := newEKSeedFromBytes(b)
+	if err != nil {
+		return s, err
+	}
+	return TeamEKSeed(seed), nil
+}
+
+func (s *TeamEKSeed) DeriveDHKey() *libkb.NaclDHKeyPair {
+	return deriveDHKey(keybase1.Bytes32(*s), libkb.DeriveReasonTeamEKEncryption)
+}
+
+type TeamEKBoxMetadata struct {
+	RecipientUID        keybase1.UID          `json:"recipient_uid"`
+	RecipientGeneration keybase1.EkGeneration `json:"recipient_generation"`
+	Box                 string                `json:"box"`
+}
+
+func postNewTeamEK(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID, sig string, boxes []TeamEKBoxMetadata) (err error) {
+	defer g.CTrace(ctx, "postNewTeamEK", func() error { return err })()
+
+	boxesJSON, err := json.Marshal(boxes)
+	if err != nil {
+		return err
+	}
+	apiArg := libkb.APIArg{
+		Endpoint:    "team/team_ek",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		NetContext:  ctx,
+		Args: libkb.HTTPArgs{
+			"team_id": libkb.S{Val: string(teamID)},
+			"sig":     libkb.S{Val: sig},
+			"boxes":   libkb.S{Val: string(boxesJSON)},
+		},
+	}
+	_, err = g.GetAPI().Post(apiArg)
+	return err
+}
+
+func publishNewTeamEK(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID, merkleRoot libkb.MerkleRoot) (metadata keybase1.TeamEkMetadata, err error) {
+	defer g.CTrace(ctx, "publishNewTeamEK", func() error { return err })()
+
+	seed, err := newTeamEphemeralSeed()
+	if err != nil {
+		return metadata, err
+	}
+
+	statement, err := fetchTeamEKStatement(ctx, g, teamID)
+	if err != nil {
+		return metadata, err
+	}
+	var generation keybase1.EkGeneration
+	if statement == nil {
+		generation = 0
+	} else {
+		generation = statement.CurrentTeamEkMetadata.Generation
+	}
+	generation++
+
+	// metadata, myTeamEKBoxed, err := signAndPublishTeamEK(ctx, g, teamID, generation, seed, merkleRoot, statment) TODO
+	metadata, _, err = signAndPublishTeamEK(ctx, g, teamID, generation, seed, merkleRoot, statement)
+
+	// TODO store result
+	//if myTeamEKBoxed == nil {
+	//	g.Log.CDebugf(ctx, "No box made for own userEK")
```

Comment:
Once the storage logic goes in, I think this should be a weird unexpected error condition. Probably worth more than a debug?